### PR TITLE
Add three analyzers and the new MyAnalysis.C template

### PR DESCRIPTION
**Pull request by kmei-cms**
Because of the implementation of JER/JEC, it is necessary to change the MyAnalysis.C template:

1) Jet.h, BJet.h, RunFisher.h now has to have arguments passed into it to determine whether to use JEC/JER up/down
2) ScaleFactors.h is ran to add the pileup/btag efficiencies into the tree to be used in analyzers
3) Make sure to cd to the right output file when writing histograms, or you will not be able to output anything

The three analyzers are:
1) CalculateBTagSF -> gives the code that calculates the BTag Efficiency as expected from Type I in the BTag recommendations. This is just so everyone has this piece of code.
2) AnalyzeBTagSF -> gives an analyzer that outputs the histograms for all three different scale factors, gives an example of how to call these scale factors
3) AnalyzeHadTrigger -> just my trigger analysis code, but I am changing this up to use NanoAOD now

### CODE REVIEW

File: Analyzer/test/Makefile [hunk 1]
```diff
@@ -126,7 +126,7 @@ $(ODIR)/%.o : $(WDIR)/%.cc
 $(ODIR)/%.o : $(WDIR)/%.cpp
 	$(CXX) $(CXXFLAGS) $(CXXDEPFLAGS)  -I$(TDIR) -I$(IFWDIR) -I$(SFWDIR) -I$(IADIR) -I$(SADIR) -I$(SSDIR) -I$(ITTDir) -I$(STTDir) -o $@ -c $<
 
-MyAnalysis: $(ODIR)/MyAnalysis.o $(ODIR)/AnalyzeBackground.o $(ODIR)/AnalyzeTopTagger.o $(ODIR)/AnalyzeEventSelection.o $(ODIR)/AnalyzeEventShape.o $(ODIR)/Analyze0Lep.o $(ODIR)/NTupleReader.o $(ODIR)/Utility.o $(ODIR)/samples.o $(ODIR)/EventShapeVariables.o $(ODIR)/bdt_350to650_fwm10_jmtev_top6.o $(ODIR)/SetUpTopTagger.o $(ODIR)/SATException.o
```
Comment (cmadrid1) [inline Analyzer/test/Makefile]: Looks like you are not compiling "AnalyzeBackground" anymore??

File: Analyzer/test/MyAnalysis.C [hunk 2]
```diff
@@ -2,19 +2,22 @@
 #include "SusyAnaTools/Tools/NTupleReader.h"
 #include "TopTagger/CfgParser/include/TTException.h"
 
-#include "Analyzer/Analyzer/include/AnalyzeBackground.h"
-#include "Analyzer/Analyzer/include/AnalyzeTopTagger.h"
-#include "Analyzer/Analyzer/include/AnalyzeEventSelection.h"
-#include "Analyzer/Analyzer/include/AnalyzeEventShape.h"
-#include "Analyzer/Analyzer/include/Analyze0Lep.h"
```
Comment (cmadrid1) [inline Analyzer/test/MyAnalysis.C]: These should not be deleted.

File: Analyzer/test/MyAnalysis.C [hunk 3]
```diff
@@ -51,12 +67,19 @@ template<typename Analyze> void run(std::set<AnaSamples::FileSummary> vvf,
         // Define classes/functions that add variables on the fly
         std::shared_ptr<RunTopTagger> rtt;
         if ( !isSkim ) rtt = std::make_shared<RunTopTagger>();
-        RunFisher runFisher;
+        RunFisher runFisher("v3",myVarSuffix);
```
Comment (cmadrid1) [inline Analyzer/test/MyAnalysis.C]: Is the RunFisher constructor able to take two arguments?

Comment (cmadrid1) [reply Analyzer/test/MyAnalysis.C]: I just checked and the current version does not. 
https://github.com/StealthStop/Framework/blob/master/Framework/include/RunFisher.h#L194

File: Analyzer/test/MyAnalysis.C [hunk 4]
```diff
@@ -51,12 +67,19 @@ template<typename Analyze> void run(std::set<AnaSamples::FileSummary> vvf,
...
+            tr.registerFunction( std::move(bTagCorrector) );
+            tr.registerFunction( std::move(pileup) );
+        }
         Muon muon;
         Electron electron;
-        Jet jet;
-        BJet bjet;
+        Jet jet(myVarSuffix);
+        BJet bjet(myVarSuffix);
+        ScaleFactors scaleFactors;
```
Comment (cmadrid1) [inline Analyzer/test/MyAnalysis.C]: Again I'm not sure if these constructors can take a variable.

File: Analyzer/test/MyAnalysis.C [hunk 5]
```diff
@@ -113,20 +138,16 @@ std::set<AnaSamples::FileSummary> setFS(const std::string& dataSets, const bool&
 int main(int argc, char *argv[])
 {
     int opt, option_index = 0;
-    bool doBackground = false, doTopTagger = false, doEventSelection = false, 
-         doEventShape = false, do0Lep = false, doStealthTT = false;
```
Comment (cmadrid1) [inline Analyzer/test/MyAnalysis.C]: Don't remove these please.

File: Analyzer/test/MyAnalysis.C [hunk 6]
```diff
@@ -113,20 +138,16 @@ std::set<AnaSamples::FileSummary> setFS(const std::string& dataSets, const bool&
...
-        {"doBackground",       no_argument, 0, 'b'},
-        {"doTopTagger",        no_argument, 0, 't'},
-        {"doEventSelection",   no_argument, 0, 's'},
-        {"doEventShape",       no_argument, 0, 'p'},
-        {"do0Lep",             no_argument, 0, 'z'},
-        {"doStealthTT",        no_argument, 0, 'x'},
+        {"calcBTagSF",         no_argument, 0, 'f'},
+        {"doBTagSF",           no_argument, 0, 'g'},
+        {"doTrigAnalysis",     no_argument, 0, 'a'},
         {"condor",             no_argument, 0, 'c'},
```
Comment (cmadrid1) [inline Analyzer/test/MyAnalysis.C]: Don't remove these either.

File: Analyzer/test/MyAnalysis.C [hunk 7]
```diff
@@ -135,16 +156,13 @@ int main(int argc, char *argv[])
...
-            case 's': doEventSelection = true;              break;
-            case 'p': doEventShape     = true;              break;
-            case 'z': do0Lep           = true;              break;
-            case 'x': doStealthTT      = true;              break;
+            case 'f': calcBTagSF       = true;              break;
+            case 'g': doBTagSF         = true;              break;
+            case 'a': doTrigAnalysis   = true;              break;
             case 'c': runOnCondor      = true;              break;
             case 'H': histFile         = optarg;            break;
             case 'D': dataSets         = optarg;            break;
```
Comment (cmadrid1) [inline Analyzer/test/MyAnalysis.C]: Same thing don't remove.

File: Analyzer/test/MyAnalysis.C [hunk 8]
```diff
@@ -159,7 +177,7 @@ int main(int argc, char *argv[])
     if(runOnCondor)
     {
         char thistFile[128];
-        sprintf(thistFile, "MyAnalysis_%s_%d.root", dataSets.c_str(), startFile);
+        sprintf(thistFile, "MyHadTrigAnalysis_%s_%d.root", dataSets.c_str(), startFile);
         histFile = thistFile;
```
Comment (cmadrid1) [inline Analyzer/test/MyAnalysis.C]: Why did you change the name of the output file name? In general it is fine but it might break some down stream code that looks for that specific name.

File: Analyzer/test/MyAnalysis.C [hunk 9]
```diff
@@ -168,37 +186,21 @@ int main(int argc, char *argv[])
...
-        {
-            printf("\n\n running Analyze0Lep\n\n") ;
-            run<Analyze0Lep>(vvf,startFile,nFiles,maxEvts,isSkim,outfile);
-        }
-        else if(doStealthTT)
-        {
-            printf("\n\n running AnalyzeStealthTopTagger\n\n") ;
-            run<AnalyzeStealthTopTagger>(vvf,startFile,nFiles,maxEvts,isSkim,outfile);
-        }
-
```
Comment (cmadrid1) [inline Analyzer/test/MyAnalysis.C]: Don't delete any of this. I just had an idea to make all of these look better that I will add after we merge everything from you.

**Comment by kmei-cms** [comment]
Newest pull request should fix the major outlying issues except for utilizing Analyze0Lep and AnalyzeStealthTopTagger, which already passes arguments into Init_Histos. Maybe we can work out a way to fix that (or find a place later in the code to Init_Histos)?

File: Analyzer/test/MyAnalysis.C [hunk 10]
```diff
@@ -68,9 +96,11 @@ template<typename Analyze> void run(std::set<AnaSamples::FileSummary> vvf,
         tr.registerFunction( std::move(runFisher) );
         tr.registerFunction( std::move(bjet) );
         tr.registerFunction( std::move(commonVariables) );
+        tr.registerFunction( std::move(scaleFactors) );
         tr.registerFunction( std::move(baseline) );
 
         // Loop over all of the events and fill histos
+        a.InitHistos(tr);
```
Comment (cmadrid1) [inline Analyzer/test/MyAnalysis.C]: Can we not InitHistos like this? Can you do it the same way that I do it in my analyzers:
https://github.com/StealthStop/Analyzer/blob/master/Analyzer/src/Analyze0Lep.cc#L384-L389

Comment (cmadrid1) [reply Analyzer/test/MyAnalysis.C]: In general I'm not sure we should or need to pass in the nTupleReader to the InitHistos like this. I think we had decided not to change the names of histogram files?

Comment (cmadrid1) [reply Analyzer/test/MyAnalysis.C]: I think how I do cuts and pass in the information for Analyze0Lep is better and will save us more time in the future. That way we are not all having to have the same type of InitHistos function.

Comment (kmei-cms) [reply Analyzer/test/MyAnalysis.C]: In general, I guess we did not need to pass the nTupleReader into the InitHistos. It was important for my B Tag Efficiency analyzer because it had to read in the sample set and sample collection in order to properly calculate the right weights needed for the b tag efficiency. I thought it might be easier to just make the change across the board.

I can change it back if that's what you want, and if you think doing InitHistos your way is better, I can adapt accordingly.

Comment (cmadrid1) [reply Analyzer/test/MyAnalysis.C]: I think it is better because we can have different InitHistos for each Analyzer and not have it in MyAnalyzer.C plus if you want you can pass in stuff like my cut map to make your life easier.

Comment (kmei-cms) [reply Analyzer/test/MyAnalysis.C]: Sounds good. I'm currently going through each analyzer and making sure it compiles.

File: Analyzer/src/Analyze0Lep.cc [hunk 11]
```diff
@@ -487,6 +487,8 @@ void Analyze0Lep::Loop(NTupleReader& tr, double weight, int maxevents, bool isQu
 
 void Analyze0Lep::WriteHistos(TFile* outfile)
 {
+    outfile->cd();
```
Comment (cmadrid1) [inline Analyzer/src/Analyze0Lep.cc]: Does this line actually do anything? I'm just curious because i'm not sure what this command does since there is only one out file.

Comment (kmei-cms) [reply Analyzer/src/Analyze0Lep.cc]: Yes, without it, none of the histograms will write out to your outfile. This is because ROOT's memory allocation ends up on the file last opened, and when we put in the ROOT files for the efficiencies (Pileup....root and allInOne...root), ROOT will try to write the histograms to those read only files and it will fail. 

If you're curious, you can run the analyzer without that line, and see what happens.

Comment (cmadrid1) [reply Analyzer/src/Analyze0Lep.cc]: Oh I see thanks.

**Comment by cmadrid1** [comment]
Here it goes. +1
